docs/rest.md

```diff
@@ -0,0 +1,623 @@
+# Snappy Ubuntu Core REST API
+
+Version: 1.0.1 DRAFT (look for "not implemented" to find bits still in flux or
+not implemented)
+
+## Versioning
+
+As the API evolves, some changes are deemed backwards-compatible (such
+as adding methods or verbs, or adding members to the returned JSON
+objects) and don't warrant an endpoint change; some changes won't be
+backwards compatible, and will be exposed under a new endpoint.
+
+## Connecting
+
+While it is expected to allow clients to connect using HTTPS over a
+TCP socket, at this point only a unix socket is supported. The socket
+is `/run/snapd.socket`.
+
+## Authentication
+
+Authentication over the unix socket is delegated to UNIX ACLs. At this point
+only root can connect to it; later, regular user access is not implemented
+yet, but should be able to once `SO_PEERCRED` is supported to determine
+privilege levels.
+
+## Responses
+
+All responses are `application/json` unless noted otherwise. There are
+three standard return types:
+
+* Standard return value
+* Background operation
+* Error
+
+Status codes follow that of HTTP.
+
+### Standard return value
+
+For a standard synchronous operation, the following JSON object is
+returned:
+
+```javascript
+{
+ "result": {},               // Extra resource/action specific data
+ "status": "OK",
+ "status_code": 200,
+ "type": "sync"
+}
+```
+
+The HTTP code must be 200.
+
+### Background operation
+
+When a request results in a background operation, the HTTP code is set
+to 202 (Accepted) and the Location HTTP header is set to the operation
+URL.
+
+The body is a json object with the following structure:
+
+```javascript
+{
+ "result": {
+   "resource": "/1.0/operations/[uuid]",     // see below
+   "status": "running",
+   "created_at": "..."                       // and other operation fields
+ },
+ "status": "Accepted",
+ "status_code": 202,
+ "type": "async"
+}
+```
+
+The response body is mostly provided as a user friendly way of seeing
+what's going on without having to pull the target operation; all
+information in the body can also be retrieved from the background
+operation URL.
+
+### Error
+
+There are various situations in which something may immediately go
+wrong, in those cases, the following return value is used:
+
+```javascript
+{
+ "result": {},            // may contain more details, for debugging
+ "status": "Bad Request", // text description of status_code
+ "status_code": 400,      // or 401, etc. (same as HTTP code)
+ "type": "error"
+}
+```
+
+HTTP code must be one of of 400, 401, 403, 404, 409, 412 or 500.
+
+### Timestamps
+
+Timestamps are presented in µs since the epoch UTC, formatted as a decimal
+string. For example, `"1234567891234567"` represents
+`2009-02-13T23:31:31.234567`.
+
+## /
```

Comment:
I'd prefer to not specify root, so the API can coexist with something else oriented towards humans if we choose to (redirects to documentation, local web site, or whatever else). We can move the listing of API versions elsewhere.